### PR TITLE
Refresh write a connector

### PR DESCRIPTION
**Pull request by jchick-bentley**
We updated the "Write a Connector" documentation to provide more code examples and explanation.

### CODE REVIEW

**Review by swwilson-bsi — APPROVED** [review]

File: docs/learning/WriteAConnector.md [hunk 1]
```diff
@@ -1,108 +1,727 @@
-# Write an iModel Connector
+---
+title: iModel Connector Developer Guide
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:2]: iModel Connector -> iTwin Connecotr

File: docs/learning/WriteAConnector.md [hunk 2]
```diff
@@ -1,108 +1,727 @@
...
+
+Bentley has authored many "domain" schemas to support connectors for many of its authoring applications. For the most aligned data (i.e., data published from your Connector uses the same classes and properties as data published from other connectors), it is best to use a domain schema.
+
+To see what domain schemas are available, see [Released Domain Schemas](ReleasedDomainSchemas.md)
+
+Sometimes BIS domain schemas are not adequate to capture all the data in the authoring application. The flow chart below can be used to assist in deciding which schema methodology to use.
+
+![Schema Methodology Decision](./SchemaDecision.PNG)
+
+#### Extending a Schema
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:211]: Did you intend for this heading to be the same level as Dynamic Schemas and Display Labels?

File: docs/learning/WriteAConnector.md [hunk 3]
```diff
@@ -1,108 +1,727 @@
...
+
+#### Dynamic Schemas
+
+When the format for incoming data in the native source is not completely known, it is not possible to map the data to a fixed schema. A solution for this scenario is called Dynamic Schema. To avoid losing data, iTwin Connector may dynamically create application-specific schemas whose classes descend from the most appropriate BIS domain classes.
+
+For example, if the native source allows for user-defined classes or properties, then as the classes and properties are read from the native source, they can be added to an iModel schema in-memory and real-time (a.k.a. dynamically). In effect, each native source file has its unique schema.
+
+As an iModel Connector always runs multiple times to keep an iModel synchronized, the schemas created by previous executions limit the schemas that subsequent executions can use. To provide consistency and enable concise changesets, the Connector adds to the previously-defined schemas (creating new schema versions). This follows the general schema update strategy defined in [Schema Versioning and Generations](https://github.com/imodeljs/imodeljs/tree/master/docs/bis/intro/schema-versioning-and-generations.md)
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:219]: iModel Connector -> iTwin Connector

File: docs/learning/WriteAConnector.md [hunk 4]
```diff
@@ -1,108 +1,727 @@
...
+
+The `DynamicSchema` custom attribute should be set on customer-specific application schemas. This custom attribute can be found in the standard schema `CoreCustomAttributes,` enabling iModelHub to detect dynamic schemas programmatically. Dynamic schemas require special handling since their name and version are typically duplicated between iModels from different work sets.
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:221]: Adding a link to the docs for CoreCustomAttributes would be helpful here.

https://www.itwinjs.org/bis/domains/corecustomattributes.ecschema/#dynamicschema

Relative link would be `../bis/domains/corecustomattributes.ecschema/#dynamicschema`

File: docs/learning/WriteAConnector.md [hunk 5]
```diff
@@ -1,108 +1,727 @@
...
+
+#### Display Labels
+
+Wherever practical, the Elements generated from an iModel Connector should be identifiable through an optimal "Display Label."
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:225]: iModel Connector -> iTwin Connector

File: docs/learning/WriteAConnector.md [hunk 6]
```diff
@@ -1,108 +1,727 @@
...
+#### Provenance and External Repository
+
+A Connector is usually dealing with two levels of provenance
+
+1. What is the identity and metadata of a file or repository synchronized into an iModel?
+2. What is the identity of the element within that repository?
+
+[RepositoryLink](https://www.itwinjs.org/reference/imodeljs-backend/elements/repositorylink/) is a specialization of a UrlLink which has a "Url" property pointing to an external resource or repository and a RepositoryGuid that identifies the external repository.
+
+[ExternalSource](https://www.itwinjs.org/reference/imodeljs-backend/elements/externalsource/) is an information container container found in a repository. A few use cases for ExternalSources are listed below:
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:273]: Duplicate `container` in the first sentence.

File: docs/learning/WriteAConnector.md [hunk 7]
```diff
@@ -1,108 +1,727 @@
...
+$npm install  @bentley/backend-itwin-client
+$npm install  @bentley/bentleyjs-core
+$npm install  @bentley/context-registry-client
+$npm install  @bentley/geometry-core
+$npm install  @bentley/imodelhub-client
+$npm install  @bentley/imodeljs-backend
+$npm install  @bentley/imodeljs-common
+$npm install  @bentley/rbac-client
+
+$npm install  --save-dev @bentley/backend-itwin-client
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:418]: A handful of these should be real deps or are duplicated from the list above.

The list should be;

```sh
$npm install  @bentley/backend-itwin-client
$npm install  @bentley/bentleyjs-core
$npm install  @bentley/context-registry-client
$npm install  @bentley/ecschema-metadata
$npm install  @bentley/geometry-core
$npm install  @bentley/imodelhub-client
$npm install  @bentley/imodeljs-backend
$npm install  @bentley/imodeljs-common
$npm install  @bentley/itwin-client
$npm install  @bentley/rbac-client
$npm install  @bentley/telemetry-client

$npm install  --save-dev @bentley/build-tools
$npm install  --save-dev @bentley/config-loader
$npm install  --save-dev @bentley/eslint-plugin
$npm install  --save-dev @bentley/oidc-signin-tool

$npm install  --save-dev chai
$npm install  --save-dev cpx
$npm install  --save-dev eslint
$npm install  --save-dev mocha
$npm install  --save-dev nyc
$npm install  --save-dev rimraf
$npm install  --save-dev typescript
```

File: docs/learning/WriteAConnector.md [hunk 8]
```diff
@@ -1,108 +1,727 @@
...
+
+- BridgeRunner: [Opens a local briefcase copy](https://github.com/imodeljs/imodeljs/tree/master/docs/learning/backend/IModelDb.md) of the iModel that is to be updated.
+- Import or Update Schema
+  - Connector: Possibly [import an appropriate BIS schema into the briefcase](https://github.com/imodeljs/imodeljs/tree/master/docs/learning/backend/SchemasAndElementsInTypeScript.md#importing-the-schema) or upgrade an existing schema.
+  - BridgeRunner: [Push](https://github.com/imodeljs/imodeljs/tree/master/docs/learning/backend/IModelDbReadwrite.md#pushing-changes-to-imodelhub) the results to the iModelServer.
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:596]: iModelServer should either be two words or changed to iModelHub

File: docs/learning/WriteAConnector.md [hunk 9]
```diff
@@ -1,108 +1,727 @@
...
+
+3. List of all physical partitions present in the iModel
+
+   ```sql
+   SELECT ECInstanceId, CodeValue from bis.physicalpartition
+   ```
+
+#### Visualizing the output
+
+To get started, please build it using the instructions provided in the [display test app README](https://github.com/imodeljs/imodeljs/blob/master/test-apps/display-test-app/README.md)
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:642]: We do not want people cloning and using the Display Test App outside of someone working directly on iTwin.js in the repo.

I'd suggest you point people to the iTwin Desktop Viewer as the way they look at their local data run through the Connector.  There is a tutorial here, https://www.itwinjs.org/learning/tutorials/develop-desktop-viewer/.

Side note: That tutorial needs to be updated, working on that today so hopefully will be in place...

File: docs/learning/WriteAConnector.md [hunk 10]
```diff
@@ -1,108 +1,727 @@
...
+Once the application is built and running, use the briefcase icon to open the output from the Connector. Some of the [frontend developer tools](https://www.npmjs.com/package/@bentley/frontend-devtools/v/1.5.0) are also useful to analyze the data. E.g. fdt inspect element.
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:643]: @pmconne do we want to, or already have, any docs on the frontend dev tools?

File: docs/learning/WriteAConnector.md [hunk 11]
```diff
@@ -1,108 +1,727 @@
...
+
+### Building a test for a Connector
+
+For examples of Connector tests, you can review the standalone and integration tests in the Connector framework. These tests use the test framework, [mocha](https://www.npmjs.com/package/mocha) and the assertion library, [chai](https://www.npmjs.com/package/chai/v/4.2.0)
+
+## Advanced Topics
+
+### Job Subjects
+
+The FrameWork will automatically create a uniquely named Subject element in the iModel. The job subject element should be a child of the root subject and must have a unique code.
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:661]: FrameWork -> Framework

File: docs/learning/WriteAConnector.md [hunk 12]
```diff
@@ -1,108 +1,727 @@
...
+
+Schemas that have been released for production use in end-user workflows evolve as new capabilities are added, and other improvements are made. To manage and track this schema evolution, schema versioning is used. See Schema Versioning and Generations for details on BIS's schema versioning strategy.
+
+### Units and Coordinate systems
+
+For the basics of coordinate systems in iModels, please see <https://www.itwinjs.org/learning/geolocation/>
+
+All coordinates and distances in an iModel must be stored in meters, and so the Connector must transform source data coordinates and distances into meters.
+
+For any iModel, a Connector should
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:678]: The lead in to the two items is weird, maybe remove the `should` and re-word it a bit or potentially re-word the list to start with "reproject the existing... " and "initializing the iModel ...".

**Comment by calebmshafer** [comment]
Awesome docs on Connectors, a much needed addition!

A couple high-level notes;

1. This should probably be backported into the `release/2.19.x` branch once it's merged into `master`. The `master` branch is currently working on 3.0 thus the docs will not be released from `master` until 3.0 is released. Backporting will be the quickest way to get them out there.
2. There should be a follow up PR to clean up a lot of the references to the "2.0" way of doing things given that the imodel-bridge package is going to be renamed along with a bunch of the types.

File: docs/learning/WriteAConnector.md [hunk 13]
```diff
@@ -1,108 +1,727 @@
...
+
+This guide section is a quick refresher on the basic concepts one needs to know to write a new Connector.
+
+#### iTwin
+
+An iTwin is an infrastructure digital twin. An iTwin incorporates different types of data repositories -- including drawings, specifications, documents, analytical models, photos, reality meshes, IoT feeds, and enterprise resource and enterprise asset management data -- into a living digital twin. This [link](http://www.bentley.com/itwin) contains additional information about iTwins and Bentley iTwin Services
+
+#### iModelHub
+
+iModelHub is for users who want their iModels hosted by Bentley in Azure. It is responsible for coordinating concurrent access to iModels and changes made to them in [ChangeSets](https://www.itwinjs.org/learning/glossary/#changeset). iModel connectors interact with iModelHub using the iModel.js API. The Connector SDK provides a framework through which a Connector can easily maintain this interaction. For more information about iModelHub, please see <https://www.itwinjs.org/learning/imodelhub/>
```
[the remark this answers is not in the document]
Comment (jchick-bentley) [reply docs/learning/WriteAConnector.md]: > Can you make the ChangeSets link a relative link to the glossary? That goes for all links on this page to itwinjs.org.

replaced links w/ learning in the path.  There are still several links w/ "reference" in the path.  It wasn't obvious to how to make those relative or at least how to test those in vs code.  I search all .md files in source tree for an example of relative path using reference , but, came up empty.

File: docs/learning/WriteAConnector.md [hunk 14]
```diff
@@ -1,108 +1,727 @@
...
+
+An iModel is a specialized information container for exchanging data associated with the lifecycle of infrastructure assets. It contains digital components assembled from many sources. They are self-describing, geometrically precise, open, portable, and secure. The file format is based on open source SQLite relational database format and forms the backbone for iTwins
+
+iModels were created to facilitate the sharing and distribution of information regardless of the source and format of the data. iModels are an essential part of the digital twin world. But a digital twin means a lot more than just an iModel.
+
+An iTwin Connector provides a workflow to easily synchronize information from various third-party design applications or data repositories into an iModel.
+
+#### iTwin Stack
+
+iTwin Stack is a platform for building Infrastructure Digital Twins on private clouds. It enables digital and system integrators to build scalable and customizable Digital Twins. It is meant for organizations unable to use iTwin Platform.
```
[the remark this answers is not in the document]
Comment (jchick-bentley) [reply docs/learning/WriteAConnector.md]: removed per e-mail discussion

File: docs/learning/WriteAConnector.md (hunk 2 above)
Comment (jchick-bentley) [reply docs/learning/WriteAConnector.md]: Changed header to be on same level as schemas

File: docs/learning/WriteAConnector.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (jchick-bentley) [reply docs/learning/WriteAConnector.md]: @calebmshafer I was able to create relative links for all the links to learning folder, but, bis and reference are still giving me some trouble and I searched my source tree for examples but I don’t see any.
e.g.
https://www.itwinjs.org/reference/imodeljs-backend/elements/repositorylink/
https://www.itwinjs.org/bis/intro/codes/
any suggestions?

File: docs/learning/WriteAConnector.md (hunk 10 above)
Comment (jchick-bentley) [reply docs/learning/WriteAConnector.md:632]: @calebmshafer I believe all comments but this and relative links are resolved.

Comment (jchick-bentley) [reply docs/learning/WriteAConnector.md]: removed last sentence.

File: docs/learning/iTwiniModelConnectorBackground.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+A complete list of available connectors can be found in [iTwin Services Community Wiki](https://communities.bentley.com/products/digital-twin-cloud-services/itwin-services/w/synchronization-wiki/47595/supported-applications-and-file-formats)
+
+Examples of iTwin Connector include:
+
+![](https://communities.bentley.com/resized-image/__size/650x450/__key/communityserver-wikis-components-files/00-00-00-05-55/Bentley.png)
+![](https://communities.bentley.com/resized-image/__size/650x450/__key/communityserver-wikis-components-files/00-00-00-05-55/3rdParty.PNG)
+
+See [Section on iTwin Synchronization](#ways-to-sync-data-to-an-itwin) for more details on existing connectors.
+
+However in certain cases, where a specific format is not covered, one can develop a new connector using  [iModel.js SDK](https://www.itwinjs.org/)
```
Comment (calebmshafer) [inline docs/learning/iTwiniModelConnectorBackground.md:43]: iModel.js SDK -> iTwin.js SDK

File: docs/learning/WriteAConnector.md [hunk 16]
```diff
@@ -1,108 +1,716 @@
...
+
+An iTwin Connector provides a workflow to easily synchronize information from various third-party design applications or data repositories into an iModel.
+
+#### Briefcases
+
+[Briefcases](./glossary#briefcase) are the local copies of iModel that users can acquire to work with the iModel. A Connector will download a briefcase locally using the BridgeRunner and change their copy of iModel. Once all the work is done, the results are then pushed into the iModel. Please see the section on [Execution sequence](#execution-sequence) on the different steps involved.
+
+#### Element
+
+iModel uses BIS schemas to describe the persistence model of the digital twin. An element represents an instance of a [bis:Element](https://prod.imodeljs.org/bis/intro/element-fundamentals/) class.
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:142]: relative link needed

File: docs/learning/WriteAConnector.md [hunk 17]
```diff
@@ -1,108 +1,716 @@
...
-  * [Push](./backend/IModelDbReadwrite.md#pushing-changes-to-imodelhub) the results to the iModelServer.
-* Convert Changed Data
-  * Connect to the data source.
-  * Detect changes to the source data.
-  * [Transform](../learning/imodel-connectors.md#data-alignment) the new or changed source data into the target BIS schema.
-  * Write the resulting BIS data to the local briefcase.
-  * Remove BIS data corresponding to deleted source data.
-  * Obtain required [Locks and Codes](./backend/ConcurrencyControl.md) from the iModel server and/or code server.
-  * [Push](./backend/IModelDbReadwrite.md#pushing-changes-to-imodelhub) changes to the iModel server.
+See this article on [Importing a schema and bootstrapping definitions](https://github.com/imodeljs/imodeljs/tree/master/docs/learning/backend/SchemasAndElementsInTypeScript.md#importing-the-schema)
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:189]: relative link needed

File: docs/learning/WriteAConnector.md [hunk 18]
```diff
@@ -1,108 +1,716 @@
...
+
+The [DynamicSchema](../bis/domains/corecustomattributes.ecschema/#dynamicschema) custom attribute should be set on customer-specific application schemas. This custom attribute can be found in the standard schema `CoreCustomAttributes,` enabling iModelHub to detect dynamic schemas programmatically. Dynamic schemas require special handling since their name and version are typically duplicated between iModels from different work sets.
+
+##### Display Labels
+
+Wherever practical, the Elements generated from an iTwin Connector should be identifiable through an optimal "Display Label."
+
+As discussed in [Element Fundamentals](https://github.com/imodeljs/imodeljs/tree/master/docs/bis/intro/element-fundamentals.md), the Display Labels are created through the following logic:
```
Comment (calebmshafer) [inline docs/learning/WriteAConnector.md:217]: relative link needed

**Review by calebmshafer — CHANGES_REQUESTED** [review]
A few missing relative links but otherwise looks good to me. Thanks for making all the updates!

File: docs/learning/iTwiniModelConnectorBackground.md (hunk 15 above)
Comment (jchick-bentley) [reply docs/learning/iTwiniModelConnectorBackground.md]: this md was deleted.  text was moved into main article earlier. it is redundant and obsolete.

**Review by calebmshafer — APPROVED** [review]

**Comment by calebmshafer** [comment]
@Mergifyio update

**Comment by calebmshafer** [comment]
@Mergifyio backport release/2.19.x

File: docs/learning/ImplementADynamicSchema.md [hunk 19]
```diff
@@ -0,0 +1,168 @@
+# Implement A Dynamic Schema
```
Comment (ColinKerr) [inline docs/learning/ImplementADynamicSchema.md:1]: Jason, did you consider extracting the code from complied examples so our build validates that the code snippets work.  Search the source for `__PUBLISH_EXTRACT_START__` to find some examples.

Also you can use `ts` as the code type for the code block, this is the standard we use in the codebase

Comment (jchick-bentley) [reply docs/learning/ImplementADynamicSchema.md:1]: @ColinKerr I was not aware of either __PUBLISH_EXTRACT_START__ nor ts.  Some of the code blocks I used came from others with no code type at all and I was seeing a markdownlint error for the missing type.  I used 'JavaScript' to make the lint error go away;  I'll replace w/ 'ts' and also look into __PUBLISH_EXTRACT_START__

**Review by williamkbentley — APPROVED** [review]
LG

File: docs/learning/ImplementADynamicSchema.md (hunk 19 above)
Comment (jchick-bentley) [reply docs/learning/ImplementADynamicSchema.md:1]: @ColinKerr just replaced ```JavaScript w ```ts  I found some clear examples of PUBLISH_EXTRACT_START and understand what is required.  In only a few places was the code taken as-is from a .ts source file.  Is this pattern a must have?

Comment (ColinKerr) [reply docs/learning/ImplementADynamicSchema.md:1]: It is important to do in the long run but it can be done as an independent PR

**Review by ColinKerr — APPROVED** [review]

**Pull request merged**